### ISSUE TITLE
[IOS] Found nil parameter when creating SKPaymentDiscount

### ISSUE THREAD

**Issue by arturomf** [j3k0/cordova-plugin-purchase#1123]
# system info
```
OSX Catalina 10.15.7
XCode 12.1
Cordova 9.0.0 // Cordova-IOS 6.1.0
Device: iPhone 11 iOS 14.1 
Plugin Version: 10.3.0
```
### Expected behavior
Init the purchase workflow after using order method

### Observed behavior
After using order method, in Xcode Console show error "Found nil parameter when creating SKPaymentDiscount". It has been working until update XCode to last version 12.1. 

All the tests has been made in real device.

### Steps to reproduce
Simply use order method in a Subscription, we are using this code:

```
try {
      await this.inAppPurchase.order(productId, {
        applicationUsername: this.user.email
      });
    } catch (error) {
      console.log('Error Ordering ' + JSON.stringify(error));
    }
```

The error is shown in xcode console but not in Safari console so it must be an error in the native code

**Comment by Yuripetusko** [comment]
Same issue here when adding Introductory offer

**Comment by reatailret** [comment]
same issue with ios 13.7 , latest XCode

**Comment by reatailret** [comment]
> 
> # system info
> 
> ```
> OSX Catalina 10.15.7
> XCode 12.1
> Cordova 9.0.0 // Cordova-IOS 6.1.0
> Device: iPhone 11 iOS 14.1 
> Plugin Version: 10.3.0
> ```
> 
> ### Expected behavior
> 
> Init the purchase workflow after using order method
> ### Observed behavior
> 
> After using order method, in Xcode Console show error "Found nil parameter when creating SKPaymentDiscount". It has been working until update XCode to last version 12.1.
> 
> All the tests has been made in real device.
> ### Steps to reproduce
> 
> Simply use order method in a Subscription, we are using this code:
> 
> ```
> try {
>       await this.inAppPurchase.order(productId, {
>         applicationUsername: this.user.email
>       });
>     } catch (error) {
>       console.log('Error Ordering ' + JSON.stringify(error));
>     }
> ```
> 
> The error is shown in xcode console but not in Safari console so it must be an error in the native code

did you solve this problem?

**Comment by aliwaqassi** [comment]
same here issue with ios 13.1

**Comment by j3k0** [comment]
Hello, this error indicates the plugin thinks you're trying to use a discount offer.

The only place `SKPaymentDiscount` is used is if the native function `purchase:` contains a 4th argument that is non-null.

For reference, this is the code from the plugin.

```objc
- (void) purchase: (CDVInvokedUrlCommand*)command {

    DLog(@"purchase: About to do IAP");
    // ...
    id discountArg = [command.arguments objectAtIndex:3];

    SKMutablePayment *payment = [SKMutablePayment paymentWithProduct:[self.products objectForKey:identifier]];
    // ...
    if ([discountArg isKindOfClass:[NSDictionary class]]) {
        NSDictionary *discount = (NSDictionary*)discountArg;
        DLog(@"purchase with discount (%@, %@, %@, %@, %@).", discount[@"id"], discount[@"key"], discount[@"nonce"], discount[@"signature"], discount[@"timestamp"]);
        if (@available(iOS 12.2, macOS 10.14.4, *)) {
            DLog(@" + discounts API available");
            payment.paymentDiscount = [[SKPaymentDiscount alloc]
              initWithIdentifier: discount[@"id"]
                   keyIdentifier: discount[@"key"]
                           nonce: [[NSUUID alloc] initWithUUIDString:discount[@"nonce"]]
                       signature: discount[@"signature"]
                       timestamp: discount[@"timestamp"]];
        }
    }
    [[SKPaymentQueue defaultQueue] addPayment:payment];
}
```

Note that if you set `store.verbosity = store.DEBUG` you should see those 2 lines in XCode:

1. `purchase: About to do IAP`
2. `purchase with discount (_some values_)`

Can you share the full logs you have in XCode when setting the verbosity to debug?

Continuing on my analysis, `purchase` is called this way from javascript:

```js
    exec('purchase', [productId, quantity, applicationUsername, discount || {}], purchaseOk, purchaseFailed);
```

I realize there's a default value set for the discount (`discount || {}`), so we might indeed try to create a discount offer with `nil` identifier, which earlier version of iOS didn't complain about. The fix is coming.

**Comment by reatailret** [comment]
I found solution. "Found nil parameter when creating SKPaymentDiscount" it is not a bug. On simualtor need use storekit configuration, [https://developer.apple.com/documentation/storekit/in-app_purchase/testing_in-app_purchases_in_xcode?language=objc](url) , its worked for me in ios 14.1 .  Another solution: check is Appstore enabled in settings, for simulator is always disabled and "connect to appstore error ..."  error in debug log. Plugin is ok.
